# Single view

In [1]:
from tqdm import tqdm
from util import *
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim import lr_scheduler
import numpy as np
import torchvision
from torchvision import datasets, models, transforms
import matplotlib.pyplot as plt
import os
import pandas as pd
import copy
from PIL import Image
%matplotlib inline

In [2]:
tr = pd.read_csv('./combine_txt/sun397_train_lt.txt', header=None, sep=' ')

class_names = {}
class_num = {}

for lab in tr[1].unique():
    temp = tr.loc[tr[1] == lab].iloc[0, 0]
    
    if 'indoor' in temp or 'outdoor' in temp:
        class_name = temp.split('/')[-3] + '/' + temp.split('/')[-2]
        class_names[lab] = class_name
        class_num[lab] = len(tr.loc[tr[1] == lab])
    else:
        class_name = temp.split('/')[-2]
        class_names[lab] = class_name
        class_num[lab] = len(tr.loc[tr[1] == lab])

In [3]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

In [4]:
#  device = torch.device('cpu')

In [5]:
LOG_DIR = './log'
DATASET = 'plain'
MODEL_ID = 'test'
DATALOADER_WORKERS = 4
LEARNING_RATE = 0.01
LR_DECAY_FACTOR = 0.1
LR_DECAY_EPOCHS = 10
DROPOUT = False
DROPOUT_RATE = 0.5
MOMENTUM = 0.9
EPOCHS = 1
BATCH_SIZE = 256
DISPLAY_STEP = 10
NUM_CLASSES = 397

if not os.path.isdir(LOG_DIR):
    os.makedirs(LOG_DIR)

In [6]:
class sun_dataset_test (torch.utils.data.Dataset):
    
    def __init__ (self, txt_file, transform=None):
        super().__init__()
        self.df = pd.read_csv(txt_file, header=None, sep=' ')
        self.transform = transform
        
    def __len__ (self):
        return len(self.df)
    
    def __getitem__ (self, idx):
        
        image = Image.open(self.df.iloc[idx, 0])
        label = self.df.iloc[idx, 1] - 1
#         image_dir = self.df.iloc[idx, 0]
        
        if self.transform:
            image = self.transform(image)
            
        return image, label

In [7]:
# transforms.RandomResizeCrop(224)
data_transform = transforms.Compose([
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ])

In [8]:
dataset = sun_dataset_test(txt_file='./sun397_test_lt.txt', transform=data_transform)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=True, num_workers=DATALOADER_WORKERS)
dataset_size = len(dataset)

In [9]:
checkpoint = torch.load('./log/epoch_save_50/epoch_34_checkpoint.pth.tar')
DROPOUT = True
DROPOUT_RATE = 0.5
FC_ADD_DIM = 4096
BIAS = True

In [10]:
# resnet = torchvision.models.resnet152(pretrained=True)
# # Freeze all layers
# for param in resnet.parameters():
#     param.requires_grad = False

resnet = torchvision.models.resnet50(pretrained=False)
    
print('Done.')    

if DROPOUT:
    print('Model using dropout.')
    resnet = MyResNet(resnet, NUM_CLASSES, DROPOUT_RATE, FC_ADD_DIM, bias=BIAS)
else:
    print('Model not using dropout.')
    # Reset the fc layer
    num_features = resnet.fc.in_features
    resnet.fc = nn.Linear(num_features, NUM_CLASSES)

# Remember to comment this out
resnet = nn.DataParallel(resnet)
    
resnet = resnet.to(device)

loss_function = nn.CrossEntropyLoss()

Done.
Model using dropout.
Bias is True
Intermediate fc dimension is: 4096
Dropout rate is: 0.500000


In [11]:
model = resnet

model.eval()

model.load_state_dict(checkpoint)

In [12]:
running_loss = 0.0
running_correct_total = 0

class_correct = torch.tensor([0. for i in range(NUM_CLASSES)])
class_total = torch.tensor([0. for i in range(NUM_CLASSES)])

# Iterate over data
for inputs, labels in tqdm(dataloader):

    inputs, labels = inputs.to(device), labels.to(device)

    # Forward
    # If on training phase, enable gradients
    with torch.set_grad_enabled(False):

        logits = model(inputs)
        _, preds = torch.max(logits, 1)
        loss = loss_function(logits, labels)
    
    # Record loss and correct predictions
    correct_tensor = (preds == labels).squeeze()
    running_loss += loss.item() * inputs.shape[0]
    running_correct_total += correct_tensor.sum().item()

    for i in range(len(labels)):
        label = labels[i]
        class_correct[label] += correct_tensor[i].item()
        class_total[label] += 1

# Epoch loss and accuracieds
test_loss = running_loss / dataset_size
test_acc_mic = running_correct_total / dataset_size
test_acc_mac = (class_correct / class_total).mean().item()

print('loss: %.3f, accuracy_micro: %.3f, accuracy_macro: %.3f' 
      % (test_loss, test_acc_mic, test_acc_mac))
    

100%|██████████| 63/63 [00:30<00:00,  2.07it/s]

loss: 4.123, accuracy_micro: 0.519, accuracy_macro: 0.519


 resnet50 epoch 34 51.9

In [16]:
class_acc.mean()

tensor(0.5193)

In [13]:
class_acc = class_correct / class_total

In [14]:
acc_df = pd.DataFrame(columns=['class_id', 'class_name', 'data_number', 'accuracy'])

In [15]:
for l in range(len(class_acc)):
    print('class id: %4d,    class name: %25s,    class data number: %5d,    accuracy: %.1f' %(l, class_names[l+1], class_num[l+1], class_acc[l]*100))
    temp_df = pd.DataFrame([[l, class_names[l+1], class_num[l+1], class_acc[l].item()*100]], columns=['class_id', 'class_name', 'data_number', 'accuracy'])
    acc_df = acc_df.append(temp_df)

class id:    0,    class name:                     abbey,    class data number:   232,    accuracy: 40.0
class id:    1,    class name:            airplane_cabin,    class data number:    12,    accuracy: 32.5
class id:    2,    class name:          airport_terminal,    class data number:   597,    accuracy: 80.0
class id:    3,    class name:                     alley,    class data number:   136,    accuracy: 82.5
class id:    4,    class name:              amphitheater,    class data number:   132,    accuracy: 67.5
class id:    5,    class name:          amusement_arcade,    class data number:    74,    accuracy: 62.5
class id:    6,    class name:            amusement_park,    class data number:   392,    accuracy: 67.5
class id:    7,    class name:          anechoic_chamber,    class data number:    46,    accuracy: 72.5
class id:    8,    class name: apartment_building/outdoor,    class data number:   259,    accuracy: 52.5
class id:    9,    class name:               apse/indo

plain_base_lr_40_ep_70_caffe_drop_0.5_dim_4096 0.451

plain_base_lr_30_ep_70_caffe_drop_0.5_dim_4096 0.448

plain_base_lr_30_ep_70_caffe_drop_0.5_dim_4096_zerobias 0.447

plain_base_lr_30_ep_70_caffe_drop_0.5_dim_2048 0.444

plain_base_lr_30_ep_70_caffe_drop_0.5_dim_1024 0.437



plain_base_lr_30_ep_70_caffe_drop_0.3_dim_4096 0.443

plain_base_lr_30_ep_70_caffe_drop_0.3_dim_2048 0.445

plain_base_lr_30_ep_70_caffe_drop_0.3_dim_1024 0.444


plain_base_lr_50_ep_150_caffe_drop_0.5_dim_4096 0.448

plain_base_lr_60_ep_150_caffe_drop_0.5_dim_4096 0.452

plain_base_lr_70_ep_150_caffe_drop_0.5_dim_4096 0.452

plain_base_lr_100_ep_150_caffe_drop_0.5_dim_4096 0.453 113



combine_base_lr_70_ep_150_caffe_drop_0.5_dim_4096 0.472 136

aug_base_lr_100_ep_150_caffe_drop_0.5_dim_4096 0.457 99

combine_aug_sm_base_lr_100_ep_150_caffe_drop_0.5_dim_4096 0.477 126

combine_aug_base_lr_100_ep_150_caffe_drop_0.5_dim_4096 0.48 110

combine_base_lr_50_ep_100_caffe_drop_0.5_dim_4096_class_aware_final 0.491

python ResNet_SUN397.py --dataset plain --epoch 150 --model_id plain_base_lr_70_ep_150_caffe_drop_0.5_dim_4096 --decay_epoch 70 --caffe_weights True --dropout True --dropout_rate 0.5 --fc_add_dim 4096